### ISSUE TITLE
ResourceWarning: unclosed socket upon driver crashes

### ISSUE THREAD

**Issue by GergelyKalmar** [SeleniumHQ/selenium#9222]
## 🐛 Bug Report

When running tests in parallel with `pytest-xdist` Chromium webdriver creation occasionally crashes or holds up a thread indefinitely for varying reasons:

```
E                 File ".../selenium-test/lib/python3.8/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
E                   raise exception_class(message, screen, stacktrace)
E               selenium.common.exceptions.SessionNotCreatedException: Message: session not created
E               from disconnected: Unable to receive message from renderer
E                 (Session info: headless chrome=88.0.4324.182)
```
```
E                 File ".../selenium-test/lib/python3.8/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
E                   raise exception_class(message, screen, stacktrace)
E               selenium.common.exceptions.WebDriverException: Message: unknown error: Chrome failed to start: crashed.
E                 (unknown error: DevToolsActivePort file doesn't exist)
E                 (The process started from chrome location /snap/chromium/1497/usr/lib/chromium-browser/chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.)
```

I've been trying to work around it by just re-trying the creation, however, I'm also getting unclosed socket warnings every time there is a crash:

```
E               During handling of the above exception, another exception occurred:
E
E               Traceback (most recent call last):
E                 File "<frozen importlib._bootstrap_external>", line 580, in _compile_bytecode
E               ResourceWarning: unclosed <socket.socket fd=19, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('127.0.0.1', 35816), raddr=('127.0.0.1', 59245)>
```

## To Reproduce

Create a `pytest` test case file called `test_case.py`:

```python
from pytest import fixture, mark
from selenium import webdriver


@fixture
def browser():
    options = webdriver.ChromeOptions()
    options.add_argument('--headless')
    options.add_argument(f'--window-size=1000,500')
    driver = webdriver.Chrome(executable_path='/snap/bin/chromium.chromedriver', options=options)
    yield driver
    driver.close()


@mark.parametrize('index', list(range(50)))
def test_create(browser, index):
    print(browser)
```

Install the following requirements into a virtual environment:
```
pytest~=6.2.2
pytest-xdist[psutil]~=2.2.1
selenium~=3.141.0
```

Add the following to `pyproject.toml`:
```
[tool.pytest.ini_options]
filterwarnings = ['error']
```

Run `pytest -n 8` a few times. Note that the errors appear only occasionally (in only a few percent of the cases), therefore you may need to change the number of parallel test runs (`-n 8`) or the number of tests (`list(range(50))`) to see it appear.

## Expected behavior

A Chromium webdriver is created correctly 100% of the times, or, at least there is no unclosed socket warnings on failed webdriver creation.

## Environment

OS: Ubuntu 20.04
Browser: Chromium
Browser version: 88.0.4324.182 (installed via snap)
Browser Driver version: chromium.chromedriver 88.0.4324.182
Language Bindings version: Python 3.141.0

**Comment by AutomatedTester** [comment]
Looking at the error messages it looks like all the problems are with Chrome itself.

**Comment by AutomatedTester** [comment]
```
E                 File ".../selenium-test/lib/python3.8/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
E                   raise exception_class(message, screen, stacktrace)
E               selenium.common.exceptions.SessionNotCreatedException: Message: session not created
E               from disconnected: Unable to receive message from renderer
E                 (Session info: headless chrome=88.0.4324.182)
```

This is Chrome Crashing

```
E                 File ".../selenium-test/lib/python3.8/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
E                   raise exception_class(message, screen, stacktrace)
E               selenium.common.exceptions.WebDriverException: Message: unknown error: Chrome failed to start: crashed.
E                 (unknown error: DevToolsActivePort file doesn't exist)
E                 (The process started from chrome location /snap/chromium/1497/usr/lib/chromium-browser/chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.)

```

Chrome Crashed.

**Comment by AutomatedTester** [comment]
Please log chromedriver issues with the chromedriver project.

https://sites.google.com/a/chromium.org/chromedriver/help

Please be sure to include a completely reproducible test script for them, without a way to reproduce the issue you are seeing there is no good way for them to fix the problem. A reproducible test script includes a WebDriver script and a link to the page in question. If you can not provide a link, create one using [jsfiddle.net](https://jsfiddle.net)

**Comment by GergelyKalmar** [comment]
Hi @AutomatedTester, thank you very much for the insight! What about the unclosed socket issue though, is that also outside of Selenium's control?

**Comment by AutomatedTester** [comment]
Truthfully it’s hard to know if this is related or not.

If we can clean up the crashes and it still occurs please raise a bug and I will look into it.

I’ve not seen this issue before so think it’s unrelated to the Python bindings or pytest-dist but there appears to be a lot of noise we need to clear before we can go into it more

**Comment by GergelyKalmar** [comment]
I have a suspicion it _is_ related to Selenium itself, because the same unclosed socket warning seems to appear with the Firefox driver too:
```
E               Traceback (most recent call last):
E                 File ".../selenium-test/lib/python3.8/site-packages/execnet/gateway_base.py", line 845, in _local_receive
E                   callback, endmarker, strconfig = self._callbacks[id]
E               KeyError: 1
E
E               During handling of the above exception, another exception occurred:
E
E               Traceback (most recent call last):
E                 File ".../selenium-test/lib/python3.8/site-packages/execnet/gateway_base.py", line 1177, in load
E                   loader(self)
E               ResourceWarning: unclosed <socket.socket fd=29, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=6, laddr=('127.0.0.1', 54216), raddr=('127.0.0.1', 56685)>
```

I simply changed the driver from Chromium to Firefox / geckodriver. As far as I understand, sockets are managed by Selenium (also seen in issue https://github.com/SeleniumHQ/selenium/issues/6878), which makes me suspect that this part needs to be solved here.

Also, although driver crashes are somewhat rare (seems to happen only in a few percent of the cases), it would be still important to deal with those as smoothly as possible to be able to lower the chance of flaky tests.

**Comment by GergelyKalmar** [comment]
I've been able to create a simpler test case which shows the issue more clearly.

Here is the test case:
```python
from selenium import webdriver

for i in range(50):
    print(f'*** Run {i}')
    options = webdriver.ChromeOptions()
    options.add_argument('--headless')
    driver = webdriver.Chrome(options=options)
    print(driver)
    driver.quit()
```
Run it as `python3 -Wd -X tracemalloc=25 test_case.py`.

Note that the unclosed socket warning does appear in some runs (in a quite flaky way).

Also note that when using `driver = webdriver.Chrome(options=options, keep_alive=False)` the issue appears every time. The problem seems to be similar to what has been discussed in https://github.com/SeleniumHQ/selenium/issues/6878 and is related to how `urllib3.PoolManager` instances are managed in the `RemoteConnection` class (they are not cleaned up properly). When applying [this patch](https://github.com/jimevans/selenium/commit/88ecc549edc5731b8cbc3210a5e4e242e1e5807c) it seems to solve the issue when `keep_alive=False` is set.

This issue appears to be fixed in the current version of https://github.com/SeleniumHQ/selenium/blob/trunk/py/selenium/webdriver/remote/remote_connection.py, so perhaps this will be resolved in `4.0`.
